go/common/sgx/pcs/quote.go

```diff
@@ -0,0 +1,584 @@
+	qh.PCESVN = binary.LittleEndian.Uint16(data[10:])
+	copy(qh.QEVendorID[:], data[12:])
+	copy(qh.UserData[:], data[28:])
+
+	qh.raw = data[:quoteHeaderLen]
+
+	return nil
+}
+
+// QEVendorID_Intel is the Quoting Enclave vendor ID for Intel.
+var QEVendorID_Intel = []byte{0x93, 0x9a, 0x72, 0x33, 0xf7, 0x9c, 0x4c, 0xa9, 0x94, 0x0a, 0x0d, 0xb3, 0x95, 0x7f, 0x06, 0x07}
+
+// SGXExtension is an ASN1 SGX extension.
+type SGXExtension struct {
+	Id    asn1.ObjectIdentifier
+	Value asn1.RawValue
+}
+
+var (
+	// PCK_SGX_Extensions is the ASN1 Object Identifier for the SGX Extensions X509 extension.
+	PCK_SGX_Extensions = asn1.ObjectIdentifier{1, 2, 840, 113741, 1, 13, 1}
+
+	// PCK_SGX_Extensions_FMSPC is the ASN1 Object Identifier for the FMSPC SGX Extension.
+	PCK_SGX_Extensions_FMSPC = asn1.ObjectIdentifier{1, 2, 840, 113741, 1, 13, 1, 4}
+
+	// PCK_SGX_Extensions_TCB is the ASN1 Object Identifier for the TCB SGX Extension.
+	PCK_SGX_Extensions_TCB = asn1.ObjectIdentifier{1, 2, 840, 113741, 1, 13, 1, 2}
+)
+
+// AttestationKeyType is the attestation key type.
+type AttestationKeyType uint16
+
+const (
+	// AttestationKeyECDSA_P256 is the ECDSA-P256 attestation key type.
+	AttestationKeyECDSA_P256 AttestationKeyType = 2
+)
+
+// String returns a string representation of the attestation key type.
+func (kt AttestationKeyType) String() string {
+	switch kt {
+	case AttestationKeyECDSA_P256:
+		return "ECDSA-P256"
+	default:
+		return fmt.Sprintf("[unknown (%x)]", uint16(kt))
+	}
+}
+
+// QuoteSignature is a quote signature.
+type QuoteSignature interface {
+	// AttestationKeyType returns the type of the attestation key used in this quote signature.
+	AttestationKeyType() AttestationKeyType
+
+	// Verify verifies the quote signature of the header and ISV report.
+	//
+	// In case of successful verification it returns the TCB level.
+	Verify(header *QuoteHeader, isvReport *ReportBody, ts time.Time, tcb *TCBBundle) (*TCBLevel, error)
+}
+
+// QuoteSignatureECDSA_P256 is an ECDSA-P256 quote signature.
+type QuoteSignatureECDSA_P256 struct {
+	Signature            SignatureECDSA_P256
+	AttestationPublicKey [64]byte
+	QEReport             ReportBody
+	QESignature          SignatureECDSA_P256
+	AuthenticationData   []byte
+	CertificationData    CertificationData
+}
+
+// AttestationKeyType returns the type of the attestation key used in this quote signature.
+func (qs *QuoteSignatureECDSA_P256) AttestationKeyType() AttestationKeyType {
+	return AttestationKeyECDSA_P256
+}
+
+// UnmarshalBinary decodes QuoteSignatureECDSA_P256 from a byte array.
+func (qs *QuoteSignatureECDSA_P256) UnmarshalBinary(data []byte) error {
+	if len(data) < quoteSigEcdsaP256MinLen {
+		return fmt.Errorf("pcs/quote: invalid ECDSA-P256 quote signature length")
+	}
+
+	copy(qs.Signature[:], data[0:])
+	copy(qs.AttestationPublicKey[:], data[64:])
+	if err := qs.QEReport.UnmarshalBinary(data[128 : 128+reportBodyLen]); err != nil {
+		return err
+	}
+	copy(qs.QESignature[:], data[512:])
```

Comment:
no `offset += len(qs.QESignature)` or the like? this section is a little bit harder than the rest to follow.